service/role.go

```diff
@@ -145,109 +215,174 @@ func (r *roleService) getRoleToken(ctx context.Context, domain, role, proxyForPr
 	return tok, nil
 }
 
-// handleExpiredHook is a handler function for gache expired hook
-func (r *roleService) handleExpiredHook(fctx context.Context, key string) {
-	domain, role := decode(key)
-	r.updateRoleToken(fctx, domain, role, "", r.expiry, r.expiry)
-}
+// refreshRoleTokenCache returns the error channel when it is updated.
+func (r *roleService) RefreshRoleTokenCache(ctx context.Context) <-chan error {
+	glg.Info("refreshRoleTokenCache started")
 
-// updateRoleToken returns RoleToken struct or error.
-// This function ask athenz to generate role token and return, or return any error when generating the role token.
-func (r *roleService) updateRoleToken(ctx context.Context, domain, role, proxyForPrincipal string, minExpiry, maxExpiry time.Duration) (*RoleToken, error) {
+	echan := make(chan error)
+	go func() {
+		defer close(echan)
 
-	tok, err, _ := r.group.Do(encode(domain, role), func() (interface{}, error) {
-		// get the role token
-		tok, err := r.token()
-		if err != nil {
-			return nil, err
-		}
+		r.domainRoleCache.Foreach(ctx, func(key string, val interface{}, exp int64) bool {
+			domain, role, principal := decode(key)
+			for err := range r.updateRoleTokenWithRetry(ctx, domain, role, principal, r.expiry, r.expiry) {
+				echan <- err
+			}
+			return true
+		})
+	}()
 
-		// concat URL string and url parameters
-		u := getRoleTokenAthenzURL(r.athenzURL, domain, role, minExpiry, maxExpiry, proxyForPrincipal)
+	return echan
+}
 
-		req, err := http.NewRequest(http.MethodGet, u, nil)
-		if err != nil {
-			return nil, err
-		}
+// handleExpiredHook is a handler function for gache expired hook
+func (r *roleService) handleExpiredHook(ctx context.Context, key string) {
+	glg.Debugf("handleExpiredHook(), key: %s", key)
+	domain, role, principal := decode(key)
+	glg.Debugf("decoded: %s, %s, %s", domain, role, principal)
+	_, err := r.updateRoleToken(ctx, domain, role, principal, r.expiry, r.expiry)
+	glg.Errorf("err: %v", err)
+}
 
-		req.Header.Set(r.athenzPrincipleHeader, tok)
+func (r *roleService) updateRoleTokenWithRetry(ctx context.Context, domain, role, proxyForPrincipal string, minExpiry, maxExpiry time.Duration) <-chan error {
+	glg.Debugf("updateRoleTokenWithRetry started, domain: %s, role: %s, proxyForPrincipal: %s, minExpiry: %s, maxExpiry: %s", domain, role, proxyForPrincipal, minExpiry, maxExpiry)
 
-		// send http request
-		res, err := r.httpClient.Do(req.WithContext(ctx))
+	echan := make(chan error, r.errRetryMaxCount)
+	go func() {
+		defer close(echan)
 
-		if err != nil {
-			return nil, err
-		}
-		defer func() {
-			if res != nil {
-				io.Copy(ioutil.Discard, res.Body)
-				if res.Body != nil {
-					res.Body.Close()
-				}
+		for i := 0; i < r.errRetryMaxCount; i++ {
```

Comment:
retry count in authorizer only count retry (1st request excluded)
https://github.com/yahoojapan/athenz-authorizer/blob/3718419a43d56e6cd73462a00fd8501c3436f221/policy/fetcher.go#L156

which 1 is better? `maxRequestCount` or `maxRetryCount`

Reply:
@kevindiu ping